File: companies/Oracle/113.path-sum-ii.go
package leetcode

// PathSum113 backtracks over root-to-leaf paths. The path slice stores the
// current route and remaining target is reduced by each node. A valid path is
// copied only at a leaf with remaining sum zero.
//
// Time: O(n*h) worst case for copying valid paths. Space: O(h) excluding output.
func PathSum113(root *TreeNode, targetSum int) [][]int {
	ans := [][]int{}
	path := []int{}
	var dfs func(*TreeNode, int)
	dfs = func(node *TreeNode, remain int) {
		if node == nil {
			return
		}
		path = append(path, node.Val)
		remain -= node.Val
		if node.Left == nil && node.Right == nil && remain == 0 {
			ans = append(ans, append([]int(nil), path...))
		} else {
			dfs(node.Left, remain)
			dfs(node.Right, remain)
		}
		path = path[:len(path)-1]
	}
	dfs(root, targetSum)
	return ans
}

File: companies/Wipro/113.path-sum-ii.java
import java.util.*;

/**
 * Algorithm:
 * Backtrack over root-to-leaf paths. The path list stores the current route,
 * and the remaining sum tracks how much is still needed. At a leaf, copy the
 * path when the remaining sum is zero.
 *
 * Java data structures:
 * ArrayList<Integer> is used for the mutable path; valid answers are copied
 * with new ArrayList<>(path).
 *
 * Complexity:
 * Time O(n * h) in the worst case due to path copies, space O(h) recursion
 * excluding output.
 */
class Solution {
    private List<List<Integer>> ans;
    private List<Integer> path;

    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        ans = new ArrayList<>();
        path = new ArrayList<>();
        dfs(root, targetSum);
        return ans;
    }

    private void dfs(TreeNode node, int remain) {
        if (node == null) {
            return;
        }
        path.add(node.val);
        remain -= node.val;
        if (node.left == null && node.right == null && remain == 0) {
            ans.add(new ArrayList<>(path));
        } else {
            dfs(node.left, remain);
            dfs(node.right, remain);
        }
        path.remove(path.size() - 1);
    }
}

